iot-hub/Samples/device/MessageReceiveSample/MessageReceiveSample.cs

```diff
@@ -19,75 +20,95 @@ namespace Microsoft.Azure.Devices.Client.Samples
-        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
-        private static readonly TimeSpan s_receiveTimeout = TimeSpan.FromSeconds(10);
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(3);
+        private readonly TimeSpan? _maxRunTime;
         private readonly DeviceClient _deviceClient;
 
-        public MessageReceiveSample(DeviceClient deviceClient)
+        public MessageReceiveSample(DeviceClient deviceClient, TimeSpan? maxRunTime)
         {
             _deviceClient = deviceClient ?? throw new ArgumentNullException(nameof(deviceClient));
+            _maxRunTime = maxRunTime;
         }
 
         public async Task RunSampleAsync()
         {
+            using var cts = _maxRunTime.HasValue
+                ? new CancellationTokenSource(_maxRunTime.Value)
+                : new CancellationTokenSource();
+            Console.CancelKeyPress += (sender, eventArgs) =>
+            {
+                eventArgs.Cancel = true;
+                cts.Cancel();
+                Console.WriteLine("Sample execution cancellation requested; will exit.");
+            };
+            Console.WriteLine($"{DateTime.Now}> Press Control+C at any time to quit the sample.");
+
             // First receive C2D messages using the polling ReceiveAsync().
-            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub for {s_receiveTimeout}...");
-            Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT Hub blade or Azure IoT Explorer to send a message to this device.");
-            await ReceiveC2dMessagesPollingAndComplete(s_receiveTimeout);
+            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub...");
+            Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT hub blade or Azure IoT Explorer to send a message to this device.");
+            await ReceiveC2dMessagesPollingAndComplete(cts.Token);
 
             // Now subscribe to receive C2D messages through a callback.
             await _deviceClient.SetReceiveMessageHandlerAsync(OnC2dMessageReceived, _deviceClient);
             Console.WriteLine($"\n{DateTime.Now}> Subscribed to receive C2D messages over callback.");
 
             // Now wait to receive C2D messages through the callback.
             // Since you are subscribed to receive messages through the callback, any call to the polling ReceiveAsync() API will now return "null".
-            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub for {s_receiveTimeout}...");
+            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub...");
             Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT Hub blade or Azure IoT Explorer to send a message to this device.");
-            await Task.Delay(s_receiveTimeout);
+
+            try
+            {
+                await Task.Delay(-1, cts.Token);
```

Comment:
by the time we reach this point cts.Token would already have been cancelled, right? this would require a new cancellation token to be created.

Reply:
ah, N is to move to the next phase, I see

Reply:
Why would it have been signaled? That only happens if we run out of time or the user presses control-c.

In this case, we're on the second part of the sample, using subscribe, but we don't have any code we're waiting on (because it is a callback).